csrc/fusion_segmenter.cpp

```diff
@@ -1437,6 +1437,62 @@ std::string toString(const SegmentedFusion* segmented_fusion) {
   return ss.str();
 }
 
+//! Sets the rfactor as root and erases rfactor of all inputs in fusion. Any
+//! non-constant expressions in those extents are replaced by new scalars with
+//! no definition. These mutations are performed throughout the Fusion so that
+//! downstream expressions dependent on the original inputs' rfactor extents can
+//! be computed properly.
+void convertInputRfactorsToRoots(Fusion* fusion) {
+  FusionGuard fg(fusion);
+
+  // Holds all Val replacements across all inputs
+  std::unordered_map<Val*, Val*> replacement_map;
+
+  for (auto tv : ir_utils::filterByType<TensorView>(fusion->inputs())) {
+    // Create a new root domain and replacement TensorDomain.
+    // Given an rfactor domain, create a new IterDomain.
+    // Otherwise, clone the previous IterDomain
+    std::vector<IterDomain*> new_root_domain;
+    auto rfactor = tv->getMaybeRFactorDomain();
+    new_root_domain.reserve(rfactor.size());
+
+    // Does the domain (root / rfactor) contain all concrete sized extents?
+    bool tv_is_concrete = true;
+    for (auto id : rfactor) {
+      if (!id->extent()->isConstScalar()) {
+        tv_is_concrete = false;
+        break;
+      }
+    }
+
+    for (const auto& id : rfactor) {
+      if (id->isRFactorProduct()) {
+        // Create new symbolic extents for rfactor iterDomains
+        auto domain_extent = (!tv_is_concrete)
```

Comment:
I think this function is fine as it's mostly just copied from tensor_view.cpp. However, I don't know why we do use all symbolic extents if `tv_is_concrete`. It seems this means that if there's any IterDomain with a symbolic extent, all IterDomains would have symbolic extents. 

This is not about this PR itself, but does it make sense?

Reply:
> Didn't we discuss duplicated replacements if members were also traversed? Would it be a concern?

This is an important point since it could lead to us concretizing one aspect of a Val but losing a previously concretized aspect (for example losing the concretized extent when we later concretize IterType). But I think we are OK if we are careful to use `maybeMutated` when creating replacements. In one case we do already do multiple mutations; we call `OptOutMutator::mutate(id)` at the beginning of our `mutate(IterDomain* id)` override.

Reply:
> It seems this means that if there's any IterDomain with a symbolic extent, all IterDomains would have symbolic extents.

Yeah that is strange. All IterDomains in the rfactor domain are checked to determine concreteness, but then the condition is only applied to rfactor products. The purpose of this function is to make the root a "standalone" domain so that we can bind input shapes to it, so it seems like all extents regardless of if they were already rfactor products should be either constant ints or pure symbolic.

We always need to create a new `IterDomain` for rfactor products in order to cut the connection to the root domain. It seems like a simpler condition would be
```c++
auto domain_extent = id->extent()->isConstScalar()
    ? id->extent()
    : IrBuilder::create<Val>(DataType::Int);
```
And in light of the above, we could move this outside of the `id->isRFactorProduct()` check, so that we would replace IDs that are not rfactor products if they have non-constant derived extents.

Reply:
Please create a separate issue for this. Let's merge this PR as is.

Reply:
Sounds good. Merging without that change. I will do that in another PR.